File: apps/api/Hrms.Application/Common/Localization/LocalizedName.cs
namespace Hrms.Application.Common.Localization;

/// <summary>
/// ชื่อ master data ที่ HR กรอกเอง (แผนก, หมวดใบแจ้ง, หัวข้อ) — คู่แฝดฝั่ง .NET ของ
/// <c>localizedName(item, locale)</c> ใน <c>packages/i18n/src/localized-name.ts</c>
///
/// <para>
/// ลำดับ fallback ตรงกับ CLAUDE.md: ภาษาที่เลือก → en → th · ค่าว่างถือว่ายังไม่ได้แปล
/// </para>
/// </summary>
public static class LocalizedName
{
    public static string For(string name, string? nameEn, string? nameId, string locale) => locale switch
    {
        "en" => First(nameEn, name),
        "id" => First(nameId, nameEn, name),
        _ => name,
    };

    /// <summary>
    /// ชื่อครบทุกภาษาที่รองรับ — ใช้ตอน queue notification เพราะยังไม่รู้ว่าผู้รับแต่ละคนอ่านภาษาอะไร
    /// (job เป็นคนเลือกทีหลังตอนส่งจริง ดู <c>NotificationPayload.LocalizedParams</c>)
    /// </summary>
    public static Dictionary<string, string> AllLocales(string name, string? nameEn, string? nameId)
        => AppLocale.Supported.ToDictionary(
            locale => locale,
            locale => For(name, nameEn, nameId, locale),
            StringComparer.Ordinal);

    private static string First(params string?[] candidates)
        => candidates.FirstOrDefault(x => !string.IsNullOrWhiteSpace(x)) ?? string.Empty;
}

File: packages/i18n/src/localized-name.ts
import type { Locale } from './locales.ts'

/**
 * master data ที่ HR กรอกเอง (ประเภทใบแจ้ง, แผนก, ตำแหน่ง, กะ ฯลฯ) เก็บชื่อ 3 ภาษาเป็นคอลัมน์
 * `name` (ไทย — ของเดิม) + `nameEn` + `nameId` (i18n Phase M) — ส่วน LeaveType/SystemRole ใช้ `nameTh` แทน `name`
 */
export type LocalizedNamed = {
  name?: string | null
  nameTh?: string | null
  nameEn?: string | null
  nameId?: string | null
}

/**
 * เลือกชื่อตามภาษาที่ผู้ใช้ตั้ง — fallback: ภาษาที่เลือก → en → th
 * ช่องที่ HR ยังไม่กรอก (null/ว่าง) ข้ามไปภาษาถัดไป จึงไม่มีทางได้ค่าว่างกลับไปถ้ายังมีชื่อไทย
 */
export function localizedName(item: LocalizedNamed | null | undefined, locale: Locale): string {
  if (!item) return ''
  const th = pick(item.nameTh) ?? pick(item.name) ?? ''
  if (locale === 'th') return th
  const en = pick(item.nameEn)
  if (locale === 'en') return en ?? th
  return pick(item.nameId) ?? en ?? th
}

function pick(value: string | null | undefined) {
  const trimmed = value?.trim()
  return trimmed ? trimmed : undefined
}
